### ISSUE TITLE
Infinite redirect when the url contains ANY "+" character breaking the application

### ISSUE THREAD

**Issue by igortrinidad** [nuxt/framework#5902]

**Comment by danielroe** [comment]
Does this only occur on vercel? Are you deploying this as a static generated site or not?

**Comment by igortrinidad** [comment]
Hi Daniel,

It's happening only in Vercel, but only if the application has pages (vue-router added)... If the application doesn't have pages the Vercel serves the content without error, so i don't believe it's something related to Vercel proxy server or build...

But what i tested too:

Build locally and serve the application (locally) using the `node .output/server/index.mjs` and the error doesn't happens...

I endup fowarding the url to my api, sanitizing it and then redirect to the application but the error remains if somebody access an url with "+" character...

So this is it, now  i'm gonna create a Vue3 project using vue-router and deploy it to Vercel to test it out if it's something related to vue-router...

**Comment by ajaydevofficial** [comment]
@danielroe Facing the same issue on Nuxt 3 SSR application deployed on google's AppEngine connected with a custom proxy server. The proxy server is receiving 302 from the Nuxt 3 Application after adding a query parameter with +

**Comment by Cosbgn** [comment]
Hi this is still happening, it's an issue because many ad platforms like Google Ads automatically add URL parameters, so you might be sending paid ads to dead pages. I've created a reproduction at:

- https://rx-gamma.vercel.app/
- https://github.com/Cosbgn/nuxt-start

Any ideas for a temporary fix? Maybe using middlewares to replace + signs with -?
Thanks a lot!

**Comment by igortrinidad** [comment]
@Cosbgn i encountered this error while i was developing google auth, so google added the "+" char on the url while redirecting back the user... My solution was redirecting to the api and then i sanitized the url back to the frontend to avoid the error, but it may not be your case if you can't control the accesed url... :/

**Comment by Cosbgn** [comment]
@igortrinidad The issue is that Google ads doesn't really like redirects, plus it would slow things down to call the api and redirect back to a static page

**Comment by igortrinidad** [comment]
@Cosbgn yeah, unfortunatelly my solution doesnt fit for your problem...

**Comment by smaeda-ks** [comment]
Hi there,

Have you tried overriding the default vue-router options? @igortrinidad 
https://v3.nuxtjs.org/guide/directory-structure/pages/#router-options

For example, if you add the following snippet in your `app/router.options.ts` file, does this change the behavior?

```ts
import type { RouterConfig } from '@nuxt/schema'

export default <RouterConfig>{
  stringifyQuery: (query) => {
    const keys = Object.keys(query)
    return keys.map(key => `${key}=${encodeURIComponent(query[key] as string)}`).join('&')
  }
}
```

**Comment by igortrinidad** [comment]
@smaeda-ks thanks a lot for your help, it solves the issue on the test repository 🙏 !

[https://test-nuxt-error.vercel.app/?xcvxcv=1231+asdasd+123](https://test-nuxt-error.vercel.app/?xcvxcv=1231+asdasd+123)

[https://github.com/igortrinidad/test-nuxt-error](https://github.com/igortrinidad/test-nuxt-error)

I'll try it for the app that i'm using Nuxt3 and give some feedback later... But I'll let the issue opened for now because it didn't solve the issue without this custom implementation...

**Comment by Cosbgn** [comment]
So this seems to happen also if you add `/` in the param, like `/?utm=fo/` - It might happen also for many other characters we don't know about.

**Comment by smaeda-ks** [comment]
This behavior is basically due to how Vercel proxy parses an incoming requested URL and how Nuxt parses the URL after that. The latter is relying on the ufo package: https://github.com/unjs/ufo

For instance,

### `/?param=foo+bar`

1. original URL: `/?param=foo+bar`
2. parsed after Vercel proxy: `/?param=foo%20bar`
3. parsed after ufo package: `/?param=foo+bar` (the effect of `encodeQueryValue()` [function](https://github.com/unjs/ufo/blob/main/src/encoding.ts#L59))
4. Nuxt returns a redirect: `location: /?param=foo+bar`

### `/?param=foo/`

1. original URL: `/?param=foo/`
2. parsed after Vercel proxy: `/?param=foo%2F`
3. parsed after ufo package: `/?param=foo/` (the effect of `parseQuery()` [function](https://github.com/unjs/ufo/blob/main/src/query.ts#L11))
4. Nuxt returns a redirect: `location: /?param=foo/`

and this causes the redirect loop. I'm not exactly sure where Nuxt is checking the parsed URL and deciding whether to return a redirect or not, but the redirect is coming from Nuxt itself, I believe.

**Comment by guyschlider** [comment]
We just had the same issue (Vercel + Nuxt3, infinite loops when a + is in the query string)

**Comment by henningko** [comment]
I'm using @smaeda-ks's workaround, but it still feels like a bug that should get attention.

**Comment by cosbgn** [comment]
I agree, I'm also using some Middleware hacks but literary every single
nuxt app deployed to Vercel is currently broken so I think it should get
some priority and it should be fixed asap.

**Comment by nethunter** [comment]
Happens to me too on Vercel, ended up replacing spaces with '-', but definitely feels like a bug.
